pkg/routeagent_driver/handlers/healthchecker/healthchecker.go

```diff
@@ -0,0 +1,331 @@
+	h.Lock()
+	defer h.Unlock()
+
+	for _, p := range h.pingers {
+		p.Stop()
+	}
+
+	h.pingers = map[string]pinger.Interface{}
+
+	close(h.stopCh)
+
+	err := h.client.Delete(context.TODO(),
+		h.localNodeName, metav1.DeleteOptions{})
+	if err != nil {
+		logger.Warningf("Error deleting RouteAgent: %s: %v", h.localNodeName, err)
+	}
+
+	return nil
+}
+
+func (h *controller) RemoteEndpointCreated(endpoint *submarinerv1.Endpoint) error {
+	if !h.config.HealthCheckerEnabled && h.State().IsOnGateway() {
+		return nil
+	}
+
+	h.remoteEndpoint[endpoint.Name] = endpoint
+
+	return h.processEndpointCreatedOrUpdated(endpoint)
+}
+
+func (h *controller) RemoteEndpointUpdated(endpoint *submarinerv1.Endpoint) error {
+	if !h.config.HealthCheckerEnabled && h.State().IsOnGateway() {
+		return nil
+	}
+
+	h.remoteEndpoint[endpoint.Name] = endpoint
+
+	return h.processEndpointCreatedOrUpdated(endpoint)
+}
+
+func (h *controller) processEndpointCreatedOrUpdated(endpoint *submarinerv1.Endpoint) error {
+	logger.Infof("Endpoint created: %#v", endpoint)
+
+	if endpoint.Spec.HealthCheckIP == "" || endpoint.Spec.CableName == "" {
+		logger.Infof("HealthCheckIP (%q) and/or CableName (%q) for Endpoint %q empty - will not monitor endpoint health",
+			endpoint.Spec.HealthCheckIP, endpoint.Spec.CableName, endpoint.Name)
+		return nil
+	}
+
+	h.Lock()
+	defer h.Unlock()
+
+	if pingerObject, found := h.pingers[endpoint.Spec.CableName]; found {
+		if pingerObject.GetIP() == endpoint.Spec.HealthCheckIP {
+			logger.Infof("HealthChecker did not change %q ", endpoint.Name)
+			return nil
+		}
+
+		logger.Infof("HealthChecker is already running for %q - stopping", endpoint.Name)
+		pingerObject.Stop()
+		delete(h.pingers, endpoint.Spec.CableName)
+	}
+
+	pingerConfig := pinger.Config{
+		IP:                 endpoint.Spec.HealthCheckIP,
+		MaxPacketLossCount: h.config.MaxPacketLossCount,
+	}
+
+	if h.config.PingInterval != 0 {
+		pingerConfig.Interval = time.Second * time.Duration(h.config.PingInterval)
+	}
+
+	if h.config.MaxPacketLossCount != 0 {
+		pingerConfig.MaxPacketLossCount = h.config.MaxPacketLossCount
+	}
+
+	newPingerFunc := h.config.NewPinger
+	if newPingerFunc == nil {
+		newPingerFunc = pinger.NewPinger
+	}
+
+	pingerObject := newPingerFunc(pingerConfig)
+	h.pingers[endpoint.Spec.CableName] = pingerObject
+	pingerObject.Start()
+
+	logger.Infof("CableEngine HealthChecker started pinger for CableName: %q with HealthCheckIP %q",
+		endpoint.Spec.CableName, endpoint.Spec.HealthCheckIP)
+
+	return nil
+}
+
+func (h *controller) RemoteEndpointDeleted(endpoint *submarinerv1.Endpoint) error {
+	h.Lock()
+	defer h.Unlock()
+
+	if pingerObject, found := h.pingers[endpoint.Spec.CableName]; found {
+		pingerObject.Stop()
+		delete(h.pingers, endpoint.Spec.CableName)
+	}
+
+	delete(h.remoteEndpoint, endpoint.Name)
+
+	return nil
+}
+
+func (h *controller) Init() error {
+	go func() {
+		wait.Until(h.syncRouteAgentStatus, RouteAgentUpdateInterval, h.stopCh)
```

Comment:
`syncRouteAgentStatus` accesses shared data so we need to lock.

```suggestion
		wait.Until( func() {
		    h.Lock()
	            defer h.Unlock()
	            
		    h.syncRouteAgentStatus()
		}, RouteAgentUpdateInterval, h.stopCh)
```

Reply:
Done.

Reply:
The `Lock()` needs to be inside the func passed to `Until` - refer to the suggested change above. You want to protect the call to `h.syncRouteAgentStatus()`.

Reply:
oh I took the comment wrongly, corrected it.